Institutional Scoping migration conf, Refs #10733

sbreker opened the pull request:
Institutional Scoping enhancement for AtoM. When activated, the feature
will allow users to be scoped to a specific repository such that browsing
and searching will restict a user to a specific repository. Repository
theming will apply to pages when scoped to give the user visual cues
indicating they are browsing with a repository. A new search block will
be present on the left hand side from which browse and search will be done.

This is the first of several commits to pull this into 2.4.x

This commit includes:
- migration 149 to set up new settings and menu items
- settings and menu item config in menus.yml and settings.yml
- Global Settings page change to add new option to "Enable insitutional
scoping"

sbreker committed:
Institutional Scoping migration conf, Refs #10733

Institutional Scoping enhancement for AtoM. When activated, the feature
will allow users to be scoped to a specific repository such that browsing
and searching will restict a user to a specific repository. Repository
theming will apply to pages when scoped to give the user visual cues
indicating they are browsing with a repository. A new search block will
be present on the left hand side from which browse and search will be done.

This is the first of several commits to pull this into 2.4.x

This commit includes:
- migration 149 to set up new settings and menu items
- settings and menu item config in menus.yml and settings.yml
- Global Settings page change to add new option to "Enable insitutional
scoping"

sbreker committed:
Institutional Block components, Refs# 10733

Institutional Scoping enhancements.

Commit includes:

- new browse menu component which provides the ability to configure custom
menu items separately from the standard browse menu. Also provides different
text labelling than the standard browse component.

- new holdings component which uses the new browse institution component.
This component is the 'Institutional Block' and is present when institutional
scoping is turned on, and the user is currently scoped to an institution (not
browsing globally). Provides search of information objects within the scoped
repository.

- custom css to support the browse institution component.

sbreker committed:
Institutional scoping helper code, Refs #10733

Institutional Scoping enhancements.

Added supporting code:

- in QubitMenu created a new alias called %currentRealm% which will contain
the current repository search context. This is used by menu items on the
Institutional Block as part of the enable_institutional_scoping feature. The
alias will be populated with search-realm if available, else it will try to
use resource->id from sf_route, else try repos param from request.

- added the method removeAttribute() to the myUser class.

sbreker committed:
Inst scoping stylesheet component, Refs #10733

This commit includes two new stylesheet components. Both components will
trigger the background theme for the repository they are part of when the
Institutional Scoping feature is turned on and the user is scoped to a repo.

The application of the Repository theme to these pages is intended to give
visual cues to the user that they are currently scoped to an institution.

These two components differ in their logic to detect when the repository
page style should be applied. Repositories always have the stylesheet applied;
the info obj module should only have it applied if the Institutional Scoping
feature is turned on, and the Achival Description plugin modules have
different display logic for the stylesheet depending on whether the Inst
Scoping feature is turned on or off.

- The information object module component will trigger the theme when the
institutional scoping setting is turned on and the search-realm is set.
This is new behaviour for the information object module and is strictly
dependent on the Institutional scoping feature being turned on.

- The sfIsadPlugin component will first look at the institutional scoping
setting and examine the search-realm for a repo. If institutional scoping
setting is off, it will fall back to the original behaviour for the Archival
Descr plugin modules: try to determine the repo from the sf_route and display
the repo stylesheet if found. This component replaces the use of the repository
stylesheet component in the Archival Description plugin modules and
incorporates new logic for the inst scoping feature.

sbreker committed:
Inst scoping stylesheet, Refs #10733

Change the following plugin modules to use the new sfIsadPlugin stylesheet
component:

- arDacsPlugin
- sfDcPlugin
- sfModsPlugin
- sfRadPlugin

MikeFE commented on lib/model/QubitMenu.php:
```diff
@@ -52,7 +52,15 @@ public function getPath($options = array())
+
+      // 'currentRealm' is used by menu items on the Institutional Block as part of the enable_institutional_scoping feature.
+      // Try using search-realm if available, else try resource->id from sf_route, else try repos param on request.
+      '%currentRealm%' => (@sfContext::getInstance()->user->getAttribute('search-realm')
```
I'd avoid both using @ and chaining ternary operators and do this before $aliases is assigned if possible. The PHP docs warn against chaining ternary ops since there are subtleties in PHP that are non-obvious when you do it, and the @ operator... I think it's better to explicitly check things with isset() and things like that.

sbreker committed:
Inst Scoping Global Settings, Refs #10733

Changes to the Global Settings controller to add the "Enable Institutional
Scoping" feature.

sbreker committed:
Inst Scoping unset search-realm, Refs #10733

This commit modfies specific controllers such that if these pages are browsed
when Institutional Scoping setting is on, and the user is currently scoped to
a repository (search-realm is set), they are assumed to be reverting to a
'global' context and will no longer be scoped to a specific institution.
In each of these circumstances, unset search-realm.

sbreker committed:
Inst Scoping search-realm exceptions,Refs #10733

These controllers have been changed to modify the state of the search-realm
user attribute.

The search-realm attribute is used to determine if a user is currently scoped
to a repository when the Insitutional Scoping setting is turned on. If
search-realm is set, the user will be scoped to this repo; if search-realm
is not set, the user will be globally scoped.

- actor browseAction.class.php: modified to set search-realm if the request
repos is set. If Institutional Scoping is ON, remove the user search-realm
attribute when request repos is not set.

- informationObject browseAction.class.php: when the request repos is NOT set
and Institutional Scoping is ON and the request 'collection' is NOT set,
remove the user search-realm attribute.

- search autocompleteAction.class.php: when the request repos is NOT set
and Institutional Scoping is ON, remove the user search-realm attribute.

- repository browseAction.class.php: When Institutional Scoping is turned
ON and if the request repos is set then set the user search-realm and add
the repo to the search query.  If repos is not set, remove the search-realm
attribute and do not add the repo to the query.

sbreker committed:
Inst Scoping infoobj templates, Refs #10733

Information Object template changes:

- 'Browse as list' link: When Institutional Scoping is ON and user is scoped
to a repo (search-realm is set), ensure 'repos' is set on the URL so that the
resulting descriptions are restricted to this repo.

- Institutional Block: When Institutional Scoping is ON and user is scoped to
a repo, display the Institutional Block (which contains the repo logo).

sbreker committed:
Inst Scoping search and repo tmplt, Refs #10733

Misc remaining changes:

- Repository _contextMenu template: modify this template to display the
Institutional Block when Institutional Scoping is ON and a repository is set.

- Global search box: When Institutional Scoping is ON, search with the top
search box is ALWAYS global. Use the Institutional Block search box to search
within an institution.

- Browse all description link: When Institutional Scoping is ON and the user
is scoped to an institution, ensure the scoped repo is part of the URL so
results are restricted to the scoped repo.

- Modify the global search box text to use the 'app_ui_label_globalSearch'
configured text.

MikeFE commented on apps/qubit/modules/actor/actions/browseAction.class.php:
```diff
@@ -81,6 +81,19 @@ public function execute($request)
+      // Add repo to the user session as realm
+      $this->context->user->setAttribute('search-realm', $request->repos);
+    }
+    elseif (sfConfig::get('app_enable_institutional_scoping'))
```
It's a nitpick but we use 'else if' when outside of templates & using the weird alternative syntax. 

elseif as all one word is only ever required in the `<?php elseif (...): ?>` style syntax.

MikeFE commented on apps/qubit/modules/search/actions/boxComponent.class.php:
```diff
@@ -21,6 +21,14 @@ class SearchBoxComponent extends sfComponent
+    // if the institutional scoping setting is on, search is always global.
+    if (sfConfig::get('app_enable_institutional_scoping'))
+    {
+      $this->repository = NULL;
```
lowercase nulls please

MikeFE commented on plugins/sfIsadPlugin/modules/sfIsadPlugin/actions/stylesheetComponent.class.php:
```diff
@@ -0,0 +1,73 @@
+
+      $resource = $request->getAttribute('sf_route')->resource;
+
+      switch (true)
```
heheh, I hadn't seen this in the wild before but I remember someone arguing for its case. It works I guess :)

sbreker committed:
Address CR recommendations, Refs #10733

- change 'elseif' to 'else if' where present
- change 'NULL' to 'null' in search/actions/boxComponent.class.php
- revised QubitMenu.class.php to remove '@' and remove nested terniary if
stmt

MikeFE commented:
:+1:
